### PR TITLE
kubeadm: component configs versioned

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

This change removes dependencies on the internal types of the kubelet and kube-proxy component configs. Along with that defaulting and validation is removed as well. kubeadm will display a warning, that it did not verify the component config upon load.

**Which issue(s) this PR fixes**:

Refs kubernetes/kubeadm#1600

**Special notes for your reviewer**:

Depends on #79220 

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-soon
/assign @neolit123
/assign @fabriziopandini 

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm no longer defaults or validates the component configs of the kubelet or kube-proxy
```

### CODE REVIEW

File: cmd/kubeadm/app/componentconfigs/registrations.go [hunk 1]
```diff
@@ -156,7 +150,9 @@ func (rs *Registrations) Default(internalcfg *kubeadmapi.ClusterConfiguration) {
 func (rs *Registrations) Validate(internalcfg *kubeadmapi.ClusterConfiguration) field.ErrorList {
 	allErrs := field.ErrorList{}
 	for kind, registration := range *rs {
-		allErrs = append(allErrs, registration.ValidateFunc(internalcfg, field.NewPath(string(kind)))...)
+		if registration.ValidateFunc != nil {
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/componentconfigs/registrations.go]: Might be I'm missing something, but I don't see where ValidateFunc are defined...

**Review by fabriziopandini — COMMENTED** [review]
@rosti thanks for this refactor
At a first pass, everything looks great, only one question to be clarified

File: cmd/kubeadm/app/componentconfigs/registrations.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: its in the registration for a component config?
https://github.com/kubernetes/kubernetes/blob/cdd186af984fd70315ecfca8dd0fd79c011a28ca/cmd/kubeadm/app/componentconfigs/registrations.go#L43

Comment (fabriziopandini) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: Yes, the field is defined there, but I can't find where this value is initialised for kubelet/kube-proxy registrations

Comment (rosti) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: They aren't ATM. I can set some funcs to do a pretty basic validation, but not the full blown, that the use of the internal types allowed us to do.
Hence, there are no validations ATM.

Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: strict unmarshal (warnings already enabled) + some basic validation should suffice.

Comment (rosti) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: Currently, only strict unmarshalling is enabled. The problem with "basic validation" is that we don't know the definition of "basic" and at this point I don't want this to degrade to blatant copy-paste of the original component config validation funcs (which are usually quite large).

**Comment by rosti** [comment]
I think, that this is enough for a first step in the right direction. Hence, we should probably merge it without adding much new stuff to it and postpone other things for later PRs.
@neolit123 @fabriziopandini WDYT?

File: cmd/kubeadm/app/componentconfigs/registrations.go [hunk 2]
```diff
@@ -96,15 +92,14 @@ var Known Registrations = map[RegistrationKind]Registration{
...
-		AddToSchemeFuncs: []AddToSchemeFunc{kubeproxyconfig.AddToScheme, kubeproxyconfigv1alpha1scheme.AddToScheme},
+		AddToSchemeFuncs: []AddToSchemeFunc{kubeproxyconfigv1alpha1.AddToScheme},
 		DefaulterFunc:    DefaultKubeProxyConfiguration,
-		ValidateFunc:     ValidateKubeProxyConfiguration,
-		EmptyValue:       &kubeproxyconfig.KubeProxyConfiguration{},
+		EmptyValue:       &kubeproxyconfigv1alpha1.KubeProxyConfiguration{},
 		GetFromInternalConfig: func(cfg *kubeadmapi.ClusterConfiguration) (runtime.Object, bool) {
 			return cfg.ComponentConfigs.KubeProxy, cfg.ComponentConfigs.KubeProxy != nil
 		},
 		SetToInternalConfig: func(obj runtime.Object, cfg *kubeadmapi.ClusterConfiguration) bool {
```
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/registrations.go]: do these methods need a rename as they are setting to public type now?

Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: hm, actually they are setting the public types in the internal kubeadm config so it's fine.

**Comment by neolit123** [comment]
/remove-priority important-soon
/priority important-soon

trying to fix the label desync.

File: cmd/kubeadm/app/componentconfigs/registrations.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/registrations.go]: i think this PR is OK and we can iterate.
/lgtm
/hold

@fabriziopandini @rosti 
feel free to hold cancel.

**Comment by fabriziopandini** [comment]
@rosti, @neolit123 
Let's keep this on hold and discuss on the office hours before merging because by switching kubeadm from the internal to the public component config APIs :  
1. We are loosing defaulting, not a problem because defaults are applied by the consumers
2. We are losing validation, and this is not ideal (errors will be much more difficult to detect for the users)
3. We are losing conversion, that is not used yet, but probably it is something we will need in the future for upgrades

This should be discussed on the Component Config team as well.
@luxas ^^ this is required for moving kubeadm out of tree

**Comment by neolit123** [comment]
@fabriziopandini 

> We are loosing defaulting, not a problem because defaults are applied by the consumers

true, it's not a problem as long as the component recognizes empty values and defaults itself.

> We are losing validation, and this is not ideal (errors will be much more difficult to detect for the users)

strict unmarshaling should provide the entry level signal for errors. the rest is up to the operator to debug on the component level. design wise, validation should have been on the side of public types.

a component should be fed **any** public type and the component itself should decide if this is a supported and valid type. not kubeadm by importing the internal logic of a component.

> We are losing conversion, that is not used yet, but probably it is something we will need in the future for upgrades

using internal types as a intermediate conversion medium is far from ideal. me and @rosti discussed the potential of public -> public converters. we can have these in kubeadm, but maybe one day in the public repos for components. i also had discussions with CAPI folks about this as they will face similar problems there with the kubeadm API.

i see the following problems in API machinery:
- defaulting should be on the side of public types.
- validation should be on the side of public types.
- conversion between public types should be possible, by a public type knowing about the previous public type.

api-machinery does not agree with components importing the internal types of another component. this is quite valid and we need to comply with that even if it require major refactors.

**Comment by fabriziopandini** [comment]
After the latest discussions, I'm ok in merging this. Ideally, the Validation func should be removed, because this is is not the case anymore, but this can be eventually done in a follow-up PR / out of the critical path for release

**Comment by rosti** [comment]
We can still have validation worked out here. As long as my KEP about the standard component config validation interface merges and we get some interface to implement.

**Comment by neolit123** [comment]
/lgtm
so that we have one day of testing before freeze at least.

File: cmd/kubeadm/app/phases/kubelet/config.go [hunk 3]
```diff
@@ -30,15 +30,15 @@ import (
...
 	"k8s.io/kubernetes/cmd/kubeadm/app/componentconfigs"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
-	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
 )
 
 // WriteConfigToDisk writes the kubelet config object down to a file
 // Used at "kubeadm init" and "kubeadm upgrade" time
-func WriteConfigToDisk(kubeletConfig *kubeletconfig.KubeletConfiguration, kubeletDir string) error {
+func WriteConfigToDisk(kubeletConfig *kubeletconfigv1beta1.KubeletConfiguration, kubeletDir string) error {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/kubelet/config.go]: similarly to the change we did in our API imports, we should have a consistent import name e.g. `kubeletconfigv1` that does not change very often.

we can do that after code freeze.

**Pull request merged**